web-client/src/client/lib.rs

```diff
@@ -680,46 +690,132 @@ impl Client {
             }
         }
 
-        let mut known_hashes = vec![];
+        let address = Address::from_any(address)?.take_native();
+        let mut known_txs = HashMap::new();
 
         if let Some(array) = known_transaction_details {
-            let plain_tx_details =
+            let known_transaction_details =
                 serde_wasm_bindgen::from_value::<Vec<PlainTransactionDetails>>(array.into())?;
-            for obj in plain_tx_details {
-                match obj.state {
-                    // Do not skip unconfirmed transactions
-                    TransactionState::New
-                    | TransactionState::Pending
-                    | TransactionState::Included => continue,
-                    _ => {
-                        known_hashes.push(Blake2bHash::from_str(&obj.transaction.transaction_hash)?)
-                    }
+            for details in known_transaction_details {
+                known_txs.insert(
+                    Blake2bHash::from_str(&details.transaction.transaction_hash)?,
+                    details,
+                );
+            }
+        }
+
+        // TODO: Get pending transactions?
+
+        let mut earliest_receipt_block_height: u32 = u32::MAX;
+
+        // Fetch transaction receipts.
+        let receipts: HashMap<_, _> = self
+            .inner
+            .consensus_proxy()
+            .request_transaction_receipts_by_address(address, min_peers, limit)
+            .await?
+            .into_iter()
+            .collect();
+
+        let mut receipts_to_fetch = vec![];
+        for (hash, block_number) in receipts.iter() {
+            if block_number < &earliest_receipt_block_height {
+                earliest_receipt_block_height = *block_number;
+            }
+
+            // Skip known transactions that are already considered confirmed.
+            if let Some(known_tx) = known_txs.get(hash) {
+                if matches!(known_tx.state, TransactionState::Confirmed)
+                    && known_tx.block_height == Some(*block_number)
+                {
+                    continue;
                 }
             }
+
+            // Ignore all receipts that are older than since_block_height.
+            if block_number < &since_block_height {
```

Comment:
Doing this here instead of in the beginning makes `earliest_receipt_block_height < since_block_height` possible. Not entirely sure if that is desirable.

Reply:
I removed `earliest_receipt_block_height`.